File: 21장_그리디_알고리즘/배낭문제.py
from turtle import back


cargo = [
  (4, 12),
  (2, 1),
  (10, 4),
  (1, 1),
  (2, 2)
]

capacity = 15

class Solution:
  def fillInBackpack(self, cargo, capacity):
    backpack_capacity = capacity
    pack = []
    money = 0

    for i in range(len(cargo)):
      pack.append([cargo[i][0] / cargo[i][1], cargo[i][0], cargo[i][1]])

    pack.sort(reverse=True)

    for i in range(len(pack)):
      if backpack_capacity - pack[i][2] >= 0:
        backpack_capacity -= pack[i][2]
        money += pack[i][1]
      else:
        remained_kg_proportion = backpack_capacity / pack[i][2]
        backpack_capacity = 0
        money += pack[i][0] * remained_kg_proportion

    return int(money)

s = Solution()

print(s.fillInBackpack(cargo, capacity))

File: 21장_그리디_알고리즘/배낭문제.js
const cargo = [
  [4, 12],
  [2, 1],
  [10, 4],
  [1, 1],
  [2, 2]
];

const capacity = 15;

function fillinBackpack(cargo, capacity) {
  let backpackCapacity = capacity;
  let money = 0;

  const pack = cargo.map(item => [item[0] / item[1], item[0], item[1]]).sort((a, b) => b[0] - a[0]);

  for (const item of pack) {
    if ((backpackCapacity - item[2]) >= 0) {
      backpackCapacity -= item[2];
      money += item[1];
    }
    else {
      const remainingWeightProportion = backpackCapacity / item[2];
      backpackCapacity = 0;
      money += remainingWeightProportion * item[0];
    }
  }

  return money;
}

console.log(fillinBackpack(cargo, capacity));